jwplatform/upload.py

```diff
@@ -0,0 +1,182 @@
+                             f"target part size.")
+
+        # Upload the parts
+        self._upload_parts(part_count)
+
+        # Mark upload as complete
+        self._mark_upload_completion()
+
+    def _upload_parts(self, part_count):
+        try:
+            filename = self._file.name
+            remaining_parts_count = part_count
+            total_page_count = part_count // MAX_PAGE_SIZE + 1
+            if total_page_count > 0:
+                for page_number in range(1, total_page_count + 1):
+                    batch_size = min(remaining_parts_count, MAX_PAGE_SIZE)
+                    page_length = MAX_PAGE_SIZE
+                    remaining_parts_count = remaining_parts_count - batch_size
+                    query_params = {'page_length': page_length, 'page': page_number}
+                    self._logger.debug(f'calling list method with page_number:{page_number} and page_length:{page_length}.')
+                    resp = self._client.list(resource_id=self._upload_id, subresource_name='parts',
+                                             query_params=query_params)
+                    body = resp.json_body
+                    upload_links = body['parts']
+                    for part_number in range(1, batch_size + 1):
+                        bytes_chunk = self._file.read(self._target_part_size)
+                        if part_number < batch_size and len(bytes_chunk) != self._target_part_size:
+                            raise IOError("Failed to read enough bytes")
+                        retry_count = 0
+                        for _ in range(self._upload_retry_count):
+                            try:
+                                self._upload_part(bytes_chunk, part_number, upload_links)
+
+                                self._logger.debug(
+                                    f"Successfully uploaded part {(page_number - 1) * MAX_PAGE_SIZE + part_number} "
+                                    f"for upload id {self._upload_id}")
+                                break
+                            except (DataIntegrityError, PartUploadError, OSError) as err:
+                                self._logger.warning(err)
+                                retry_count = retry_count + 1
+                                self._logger.warning(
+                                    f"Encountered error upload part {(page_number - 1) * MAX_PAGE_SIZE + part_number} "
+                                    f"of {part_count} for file {filename}.")
+                                if retry_count >= self._upload_retry_count:
+                                    self._file.seek(0, 0)
+                                    raise MaxRetriesExceededError(
+                                        f"Max retries ({self._upload_retry_count}) exceeded while uploading part"
+                                        f" {part_number} of {part_count} for file {filename}.", err)
```

Comment:
```suggestion
                                    raise MaxRetriesExceededError(
                                        f"Max retries ({self._upload_retry_count}) exceeded while uploading part"
                                        f" {part_number} of {part_count} for file {filename}.") from err
```
It might be better to raise this error from another

Reply:
Done